src/Runtime/Core/Collections/ReadOnlyArray.cs

```diff
@@ -0,0 +1,193 @@
+    /// <see cref="ReadOnlyArray{T}"/> doesn't have space overhead compared to regular arrays.
+    /// </remarks>
+    [Serializable]
+    [DataContract]
+    public struct ReadOnlyArray<T> : IReadOnlyList<T>
+    {
+        /// <summary>
+        /// Regular array that holds data.
+        /// </summary>
+        [DataMember]
+        private readonly T[] array;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ReadOnlyArray{T}"/> structure.
+        /// </summary>
+        [Construction("CloneArray")]
+        public ReadOnlyArray(T[] array)
+        {
+            this.array = array;
+        }
+
+        /// <summary>
+        /// Gets a boolean value which is true if this ReadOnlyArray wraps null array.
+        /// </summary>
+        public bool IsNull => this.array == null;
+
+        /// <inheritdoc/>
+        public T this[int index] => this.array[index];
+
+        /// <inheritdoc/>
+        public int Count => this.array.Length;
+
+        /// <summary>
+        /// Creates a mutable copy of this array.
+        /// </summary>
+        public T[] CloneArray() => (T[])this.array.Clone();
+
+        /// <summary>
+        /// Returns enumerator over elements of array.
+        /// </summary>
+        /// <remarks>
+        /// This is value-type non-virtual version of enumerator that is used by compiler in foreach loops.
+        /// </remarks>
+        public ReadOnlySegmentEnumerator<T> GetEnumerator() =>
+            new ReadOnlySegmentEnumerator<T>(this, 0, this.array.Length);
+
+        /// <inheritdoc/>
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() =>
+            new ReadOnlySegmentEnumerator<T>(this, 0, this.array.Length);
+
+        /// <inheritdoc/>
+        IEnumerator IEnumerable.GetEnumerator() =>
+            new ReadOnlySegmentEnumerator<T>(this, 0, this.array.Length);
+
+        /// <summary>
+        /// Helper method which allows to cast regular arrays to read only versions implicitly.
+        /// </summary>
+        public static implicit operator ReadOnlyArray<T>(T[] array) => new ReadOnlyArray<T>(array);
+    }
+
+    /// <summary>
+    /// A version if <see cref="ArraySegment{T}"/> which can not be mutated.
+    /// </summary>
+    public struct ReadOnlyArraySegment<T> : IReadOnlyList<T>
+    {
+        /// <summary>
+        /// Underlying read-only array.
+        /// </summary>
+        private readonly ReadOnlyArray<T> array;
+
+        /// <summary>
+        /// Index of the first element which belongs to this segment.
+        /// </summary>
+        private readonly int begin;
+
+        /// <summary>
+        /// Index of the first element which does not belong to this segment.
+        /// </summary>
+        private readonly int end;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ReadOnlyArraySegment{T}"/> structure.
+        /// </summary>
+        public ReadOnlyArraySegment(ReadOnlyArray<T> array, int begin, int end)
```

Comment:
Given that this constructor is public, I would suggest some basic error checking be applied. Some simple null and bounds checking would suffice.

Reply:
Added